File: Cubism3-ARKit/MenuViewController.m
//
//  MenuViewController.m
//  Cubism3-ARKit
//
//  Created by Yuma Matsune on 2018/08/17.
//  Copyright © 2018 matsune. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "MenuViewController.h"
#import "MenuCell.h"

@interface MenuViewController()

@property (nonatomic) NSArray *titles;
@property (nonatomic) NSArray *images;

@end

@implementation MenuViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.titles = [[NSArray alloc] initWithObjects:@"Restart", nil];
    self.images = [[NSArray alloc] initWithObjects:[UIImage imageNamed:@"restart"], nil];
}

- (void)viewWillLayoutSubviews {
    self.preferredContentSize = CGSizeMake(200, 70);
}

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.titles.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    MenuCell *cell = (MenuCell*)[collectionView dequeueReusableCellWithReuseIdentifier:@"MenuCell" forIndexPath:indexPath];
    [cell bindImage:self.images[indexPath.row] Title:self.titles[indexPath.row]];
    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row == 0) {
        if ([self.menuDelegate respondsToSelector:@selector(didSelectRestart)]) {
            [self.menuDelegate didSelectRestart];
        }
    }
}

@end

File: Cubism3-ARKit/menu/MenuViewController.swift
//
//  MenuViewController.swift
//  Cubism3-ARKit
//
//  Created by Yuma Matsune on 2018/08/18.
//  Copyright © 2018 matsune. All rights reserved.
//

import Foundation
import UIKit

struct MenuItem {
    let title: String
    let image: UIImage
}

protocol MenuViewControllerDelegate: class {
    func didSelectRestart()
    func didChange(isHiddenFace: Bool)
    func didSelctChangeAvatar()
}

final class MenuViewController: UICollectionViewController {
    
    public var isHiddenFace = false
    public weak var delegate: MenuViewControllerDelegate?
    
    private let items: [MenuItem] = [MenuItem(title: "Restart", image: #imageLiteral(resourceName: "restart")),
                                     MenuItem(title: "Face", image: #imageLiteral(resourceName: "none")),
                                     MenuItem(title: "Change", image: #imageLiteral(resourceName: "avatar"))]
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        preferredContentSize = CGSize(width: 200, height: 70)
    }
    
    override func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }
    
    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return items.count
    }
    
    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "MenuCell", for: indexPath) as? MenuCell else {
            fatalError("could not dequeue menuCell")
        }
        let menu = items[indexPath.row]
        cell.bind(menu: menu)
        if indexPath.row == 1 {
            cell.isSelected = isHiddenFace
        }
        return cell
    }
    
    override func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        if indexPath.row == 0 {
            delegate?.didSelectRestart()
        } else if indexPath.row == 1 {
            delegate?.didChange(isHiddenFace: !isHiddenFace)
        } else {
            delegate?.didSelctChangeAvatar()
        }
        dismiss(animated: true, completion: nil)
    }
}
